File: spring_security_jwt/src/main/java/com/revy/authapp/domain/common/BaseEntity.java
package com.revy.authapp.domain.common;

import jakarta.persistence.Column;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import lombok.Getter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;


@Getter
@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
public abstract class BaseEntity<T> {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name="id")
    private T id;

//    @CreatedBy
//    @Column(name = "created_by", nullable = false, length = 50, updatable = false)
//    private String createdBy;

    @CreatedDate
    @Column(name = "created_date", updatable = false)
    private LocalDateTime createdDate = LocalDateTime.now();

//    @LastModifiedBy
//    @Column(name = "last_modified_by", length = 50)
//    private String lastModifiedBy;

    @LastModifiedDate
    @Column(name = "last_modified_date")
    private LocalDateTime lastModifiedDate = LocalDateTime.now();
}

File: _kotlin/example_kotlin/src/main/kotlin/com/revy/domain/entity/common/BaseEntity.kt
package com.revy.domain.entity.common

import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass
import org.hibernate.Hibernate
import java.util.UUID

@MappedSuperclass
abstract class BaseEntity {
    @Id
     val id: UUID? = null

    final override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false
        // 프록시 대응: 실제 클래스 비교
        val thisClass = Hibernate.getClass(this)
        val otherClass = Hibernate.getClass(other)
        if (thisClass != otherClass) return false
        other as BaseEntity
        return id != null && id == other.id
    }

    // ★ 상수 반환: 영속화 전(id=null) → 영속화 후(id 할당) 해시가 바뀌면
    //   HashSet/HashMap에서 엔티티를 잃어버림
    final override fun hashCode(): Int = Hibernate.getClass(this).hashCode()
}